lib/modules.nix

```diff
@@ -782,11 +783,22 @@ rec {
     defsFinal = defsFinal'.values;
 
     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
-    mergedValue =
+    mergedValue = _mergedValueWithContext "unspecified";
+
+    # internal, unstable interface, only used locally
+    _mergedValueWithContext = parentTypeName:
       if isDefined then
         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal
         else let allInvalid = filter (def: ! type.check def.value) defsFinal;
-        in throw "A definition for option `${showOption loc}' is not of type `${type.description}'. Definition values:${showDefs allInvalid}"
+        in if parentTypeName == "listOf" && type.name != "functionTo" && any (def: isFunction def.value) defsFinal
+        then throw ''
+          A list element definition for option `${showOption (init loc)}' is not of type `${type.description}', but a function. Definition values:${showDefs allInvalid}
+          Did you mean to apply the function to the next element instead? Add parenthesis to do so, e.g. like this:
+              ${showOption (init loc)} = [
+                (neovim.override { vimAlias = true; })
```

Comment:
This will be confusing when there's an actual `neovim` in the user configuration.

An abstract example may be better, and it may also be helpful to give a "before and after" for those who can't conceive of such a weird list syntax to exist.

```
Wrong:
  [ foo.override { bar = true; } ]
Correct:
  [ (foo.override { bar = true; }) ]
```

Mixing user input and examples will also raise the less-than-useful questions: why does it suggest to change the value for <option> something entirely different? Am I supposed to change my config to use a path-style attribute definition; I though those were equivalent?
So I think it's better to keep it simple and only report an example as part of the explanation and let the user infer how it applies to their config.

Nitpick: the plural of parenthesis is parentheses.